src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -744,57 +596,48 @@ public SolutionState WithProjectFilePath(ProjectId projectId, string? filePath)
         /// Create a new solution instance with the project specified updated to have
         /// the specified compilation options.
         /// </summary>
-        public SolutionState WithProjectCompilationOptions(ProjectId projectId, CompilationOptions options)
+        public (SolutionState, ProjectState newProject) WithProjectCompilationOptions(ProjectId projectId, CompilationOptions options)
         {
             var oldProject = GetRequiredProjectState(projectId);
             var newProject = oldProject.WithCompilationOptions(options);
 
             if (oldProject == newProject)
             {
-                return this;
+                return (this, oldProject);
             }
 
-            return ForkProject(newProject, new CompilationAndGeneratorDriverTranslationAction.ProjectCompilationOptionsAction(newProject, isAnalyzerConfigChange: false));
+            return ForkProject(newProject);
         }
 
         /// <summary>
         /// Create a new solution instance with the project specified updated to have
         /// the specified parse options.
         /// </summary>
-        public SolutionState WithProjectParseOptions(ProjectId projectId, ParseOptions options)
+        public (SolutionState, ProjectState newProject) WithProjectParseOptions(ProjectId projectId, ParseOptions options)
         {
             var oldProject = GetRequiredProjectState(projectId);
             var newProject = oldProject.WithParseOptions(options);
 
             if (oldProject == newProject)
             {
-                return this;
+                return (this, oldProject);
             }
 
-            if (this.PartialSemanticsEnabled)
-            {
-                // don't fork tracker with queued action since access via partial semantics can become inconsistent (throw).
-                // Since changing options is rare event, it is okay to start compilation building from scratch.
-                return ForkProject(newProject, forkTracker: false);
-            }
-            else
-            {
-                return ForkProject(newProject, new CompilationAndGeneratorDriverTranslationAction.ReplaceAllSyntaxTreesAction(newProject, isParseOptionChange: true));
-            }
+            return ForkProject(newProject);
```

Comment:
the concepts of Partial-Semantics, fork-tracker, and the translation-actions are all part of SolutionCompilationState and this logic moved there.